### ISSUE TITLE
Error during exposure logging / EN Code: 16

### ISSUE THREAD

**Issue by tibor** [corona-warn-app/cwa-app-ios#1000]
I got an alert with the following message today:
An error occured during exposure logging. EN Code: 16

Tapping on „Learn more“ wouldn’t do anything except closing the alert popup.

![IMG_6391](https://user-images.githubusercontent.com/3136416/89286573-c2146400-d652-11ea-8e7d-221bc389f6dd.PNG)

Is there anything I would need to do as a user?

---
Internal Tracking ID: EXPOSUREAPP-2018

**Comment by Ein-Tim** [comment]
Which version of iOS are you using?

**Comment by tibor** [comment]
iOS 13.6 and Corona Warn App 1.1.2

**Comment by Ein-Tim** [comment]
@GPclips could you help here please?

**Comment by GPclips** [comment]
Hello @tibor,

thanks for sharing. I will contact the development team and label it as a bug, to make sure this will be handled as soon as possible.
Meanwhile, I like to ask you to explain the steps to reproduce the issue, even if it is just that you have opened the app.

@Ein-Tim thanks for your great support!

Best Regards,
LMM

Corona-Warn-App Open Source Team

**Comment by tibor** [comment]
Steps to reproduce: I'm not entirely sure, but I think I opened a push notification from the app and got this in app alert. It's either this or I manually opened the app.

**Comment by haosap** [comment]
Hi @tibor ,
Thanks for your report. Does the Error 16 occurs every time, once you open the app? Or only occurred once..

**Comment by tibor** [comment]
@haosap for now only once

**Comment by haosap** [comment]
If it only occurs once, I think it is pretty hard to reproduce.

**Comment by yannikbloscheck** [comment]
I had the same issue today. I just manually opened the app. It apparently also just refreshed. So there might be a relation there.
The second bug is that tapping "Learn more" just closes the alert and nothing more. This actually might scare and confuse normal users. Nobody wants to hear that "something went wrong" with a sensitive health thing and then the "Learn more" button doesn't even work.
It occurred on iOS 14 Developer Beta 4 with the app version 1.1.2. It just occurred once, but, if I remember, I will open the app tomorrow evening again around the same time when the daily refresh should happen and check if it does happen again.

**Comment by yannikbloscheck** [comment]
Couldn't reproduce it today, but given the error explanation in the Apple documentation (if I guessed right and it's the actually framework error code) it should not be forwarded to the user, but simply handled by the app (for example retrying after a few seconds).

**Comment by Tho-Mat** [comment]
Got the same:
![2020-09-01 at 06 44](https://user-images.githubusercontent.com/67630959/91858834-01da6580-ec6a-11ea-9704-05a40a7e653a.jpeg)
If i look in the exposure log it shows the same time.

**Comment by Ben-WD** [comment]
Seeing the same alert from time to time when opening the app on OS 14 dev beta 6.

**Comment by DerLobi** [comment]
Got it after opening the app (I don't open the app regularly, only every now and then).
Tapping on "Find out more" does nothing.
Can not reproduce when killing and reopening the app.

The UI shows "Updated at 11:02", the same time as when I opened the app, so the error could maybe be related to the sync.

![IMG_2709](https://user-images.githubusercontent.com/1935447/92095149-97463880-edd5-11ea-86b4-ea6cef1fdfa2.PNG)

**Comment by Ben305** [comment]
Seeing exactly the same @DerLobi

**Comment by sin-azucar** [comment]
I experienced this error as well a few days ago. It seemed to have no effect, though. The app updated itself as I tapped "Find out more" which is a no-op, as others have mentioned.

Even if it's not easily reproducable, I think we can agree that it can be handled better:

- The fact that tapping "Find out more" does nothing is definitely a bug
- According to Apple, error 16 means:
> ENError.Code.dataInaccessible
> The user must unlock the device before it can access data.

(source: https://developer.apple.com/documentation/exposurenotification/enerror/code/datainaccessible)

It seems strange to me that the device must be unlocked for Exposure Detection to take place (this again defeats the functionality of the background job).

IMHO this should be clarified with Apple. In any case it's not helpful displaying this error to the user who - at this point - obviously has unlocked the device already. 😉

**Comment by git4me8you** [comment]
Hi, same issue here - iPhone SE (old) - latest iOS, latest corona warn app - error occurred under 13.6.1 and re-occurred after ios update to 13.7 and seems to be correlated with massive battery consumption (draining 25%+ of remaining battery charge over night in both cases). Might be stuck at attempt to connect for data update over night, supported by observation that server data was only updated after actually accessing application. Not observed, yet, with iPhone SE 2020.

**Comment by svengabr** [comment]
Thank you, everyone, for the additional feedback. The internal Jira ticket was updated with your additional information.

Best regards,
SG

Corona-Warn-App Open Source Team

**Comment by kaibader** [comment]
Seems like I just code got the same error message in my App a few minutes ago...
iOS 13.7 (iPhone Xs); App Version 1.3.0 (6)
I tried clicking onto "Mehr Erfahren" but nothing happened and the message didn't reappear so far.

Best regards,
Kai

![Code16](https://user-images.githubusercontent.com/1196957/93170787-2cc3bf80-f728-11ea-9038-14af86eb7f1d.png)

**Comment by m33x** [comment]
iOS 13.7 CWA 1.3.0

I opened the app manually (as I do every day) and I got this. It’s not the first time either.

 ![8A2B622C-42BD-4A64-85C9-312D28FFED9C](https://user-images.githubusercontent.com/883685/94242717-c18f9f80-ff16-11ea-877f-4f2e133dea2e.png)

**Comment by m-rossi** [comment]
I also got this error today after opening the app.
iPhone X, iOS 14.0.1, CWA 1.3.2

![IMG_2559](https://user-images.githubusercontent.com/4509134/94989459-3e54f600-0575-11eb-9b6b-26db612a96dd.PNG)

The app refreshed the same moment that error appeared.

**Comment by sfaxiaberlin** [comment]
Hi, I had the same error message today and want to know what it means please. Using iPhone 8 iOS 14.0.1
![4A8D7B05-896B-4CA2-A1E9-42C8DF9B62B3](https://user-images.githubusercontent.com/72817724/95894694-bdbea280-0d89-11eb-8a10-ce986400f403.png)

**Comment by ndegendogo** [comment]
@sfaxiaberlin do you remember the context when this error message popped up? Did you just open the app before? Did you return from a submenu? Or was the app just open and you were looking at the status? Do you remember if the screen was locked?

And: could you please give your cwa version as well? Thanks.

**Comment by sfaxiaberlin** [comment]
Version 1.3.2 (0)
I opened the App and the error occured. It was not possible to click "Mehr erfahren". I open the App every day to check my amount of contacts with no risk

**Comment by ndegendogo** [comment]
*Disclaimer: I am not one of the project maintaimers here*

@sfaxiaberlin Thanks.
The meaning of EN error 16 is explained [here](https://github.com/corona-warn-app/cwa-app-ios/issues/1000#issuecomment-686443380), including a link to the original iOS documentation.
To me it looks like a race condition during start of the app. As you open the app daily, but don't get this message regularly, I would not worry too much. Of course, such a message is not nice, and the app should give more explanations to the user (or retry silently, as was already suggested). On the other hand, this is not the highest-prio issue.

Maybe you could investigate one more detail, and check the timestamps of your checks in the ENF log file. Was the time of this error message preceded by any unexpected pattern, like overly long delay of the check?

**Comment by Bergruebe** [comment]
I have the same problem, Background update is activated, just doesn't work.
Something about twice a week I get a Push-Notification from the app, to check my Status. Then I get the error ( `Fehler [...] EN Code:16`).
I use `iOS 14.2 (18B92) | CWA Version 1.5.3 (1)`

**Comment by Ein-Tim** [comment]
@Bergruebe 
But is the App working again after a restart or waiting for 24h or are you not able to use it at all?

**Comment by Bergruebe** [comment]
Yes it works, but not automatically. The Check is only if I open the App, not in the night automatically.

**Comment by Ein-Tim** [comment]
Okay @Bergruebe 

Nice to hear that it works again, the Background check Issue is tracked here: #916 & #1227

**Comment by ndegendogo** [comment]
> about twice a week I get a Push-Notification from the app, to check my Status. Then I get the error ( Fehler [...] EN Code:16).

@Bergruebe could you please post a screenshot of the push notification and of the error message here?

> Background update is activated, just doesn't work

As @Ein-Tim already suggested. Please post a comment on this at #916 , so that the devs can pick it up and do a more detailed analysis.

**Comment by Nungu** [comment]
I got a similar error message when starting the app today. The "Find out more" option shows nothing. After this error the app worked as usual.

iPhone 12, IOS 14.2.1, App 1.7.1

Thanks,
  Nungu

![Screenshot 2020-11-30 at 09 01 04](https://user-images.githubusercontent.com/24532592/100670373-0db7ef80-335f-11eb-8b73-c85dc3c747e3.png)

**Comment by ndegendogo** [comment]
This ticket is now on hold since nearly 4 months ... but the issue seems to occur again and again....

I understand that this is not 100% reproducible; but several people in this thread have already analysed the meaning of this error code 16, and also given suggestions how it could be handled. See [here](https://github.com/corona-warn-app/cwa-app-ios/issues/1000#issuecomment-670966214) and [here](https://github.com/corona-warn-app/cwa-app-ios/issues/1000#issuecomment-686443380).

So maybe you could plan to try to fix this?

**Comment by Ein-Tim** [comment]
@ndegendogo 

I'm totally on your side, if I understand everything correctly here, "fixing" means to just hide this Issue from the User, or?

**Comment by ndegendogo** [comment]
Well - it seems to be triggered from a (failed) risk check. So maybe it is worth a retry of this check after a short waiting time.

I am just guessing now, but maybe the root cause is a race condition at start of the app, and the risk check is sometimes started too early.

**Comment by Ein-Tim** [comment]
@ndegendogo 

Yes, I think a retry is already triggered (https://github.com/corona-warn-app/cwa-app-ios/issues/1000#issuecomment-670672931 & https://github.com/corona-warn-app/cwa-app-ios/issues/1000#issuecomment-703083387) or this Error does not have influence on the already running check. 

So I think, a very simple Solution would be to hide this Error-Message from the user. 

I would love to take a look at the EN-Log from somebody who experienced this Issue.

Maybe @Nungu, could you please upload your EN-Log here? (How to see below) This log does not contain any private data, but the number of encounters you had. 


<details>
<summary>How to export Exposure Notificaiton Logs under iOS 13.7 or newer</summary>

**German:**

`Einstellungen` ➡️ `Begegnungsmitteilungen` ➡️ `Status von Begegnungsaufzeichnungen` ➡️ `Begegnungsüberprüfungen` ➡️ `Überprüfungen exportieren`

**English:**

`Settings` ➡️ `Exposure Notifications` ➡️ `Exosure Logging Status` ➡️ `Exposure Checks` ➡️ `Export Exposure Checks`

**To upload the Log here, just rename the file from .json to .txt**

</details>

**Comment by dsarkar** [comment]
Dear @ndegendogo, @Ein-Tim, @Nungu,

Many thanks for the input from your side. We will approach the developers now. 

Best wishes,
DS

---
Corona-Warn-App Open Source Team

**Comment by Nungu** [comment]
Hi @Ein-Tim ,

here comes the log file.

I think an error message should be somehow actionable. If it's already known that this error can be ignored, then there is no need to expose it to the user, ihmo. And for this target audience "Find out more" shouldn't be just empty.

Cheers,
  Nungu

[ExposureChecks-2020-12-01.txt](https://github.com/corona-warn-app/cwa-app-ios/files/5620971/ExposureChecks-2020-12-01.txt)

**Comment by ndegendogo** [comment]
> for this target audience "Find out more" shouldn't be just empty

@Nungu fully agreed. Feel free to upvote #1353

**Comment by ndegendogo** [comment]
Thanks a lot @dsarkar

**Comment by dsarkar** [comment]
FYI, PR  https://github.com/corona-warn-app/cwa-app-ios/pull/1608.

Best wishes,
DS

---
Corona-Warn-App Open Source Team

**Comment by Ein-Tim** [comment]
Although this will be fixed by #1608, I will quickly tell you what I found in the Log file. @Nungu

The log does not contain any checks which were stopped, so I think the EN Code: 16 doesn't even have any influence on a Risk-Check. 
So hiding this Error is the best solution 👍

**Comment by ndegendogo** [comment]
Wow - this was fast! 😀😀😀
Thanks again, @dsarkar.

**Comment by dsarkar** [comment]
@ndegendogo, thanks! We will forward this also to the developers! Best wishes,
DS

---
Corona-Warn-App Open Source Team

**Comment by techniker** [comment]
Same error here. "Find out more" only closes the dialog.

**Comment by dsarkar** [comment]
Hi @techniker, Thanks for reporting. This problem is resolved and the solution will be implemented in a future release. 
Best wishes,
DS

---
Corona-Warn-App Open Source Team

**Comment by techniker** [comment]
> Hi @techniker, Thanks for reporting. This problem is resolved and the solution will be implemented in a future release. 
> 
> Best wishes,
> 
> DS
> 
> 
> 
> ---
> 
> Corona-Warn-App Open Source Team

Thanks! :)

**Comment by dsarkar** [comment]
@techniker, this should solve it https://github.com/corona-warn-app/cwa-app-ios/issues/1000#issuecomment-736485835.

**Comment by Ein-Tim** [comment]
Fixed with version 1.9.1.

**Comment by dsarkar** [comment]
Dear community, 

Thank you for all your contributions. As @Ein-Tim says, this issue is fixed now. Thanks again! Best wishes, DS

---
Corona-Warn-App Open Source Team
